### PR TITLE
Added datasetBlob at CreateDatasetVersion request & Fix delete datasetVersion logic

### PR DESCRIPTION
**Pull request by anandjakhaniya**

### CODE REVIEW

File: backend/src/main/java/ai/verta/modeldb/cron_jobs/DeleteEntitiesCron.java [hunk 1]
```diff
@@ -630,6 +630,7 @@ private void deleteRepositories(Session session) {
                               ModelDBConstants.DEFAULT_VERSIONING_BLOB_LOCATION));
                   deleteLabels(session, compositeId, IDTypeEnum.IDType.VERSIONING_REPO_COMMIT_BLOB);
                   deleteAttribute(session, compositeId);
+                  deleteTagEntities(session, repository.getId(), commitEntity.getCommit_hash());
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/cron_jobs/DeleteEntitiesCron.java]: Please extract this in different PR , this is a fix worthy to be in master

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java [hunk 2]
```diff
@@ -382,18 +390,122 @@ public CommitEntity getCommitEntity(
...
+        Query<CommitEntity> getCommitQuery =
+            session.createQuery(
+                "From "
+                    + CommitEntity.class.getSimpleName()
+                    + " c WHERE c.commit_hash = :commitHash",
+                CommitEntity.class);
+        getCommitQuery.setParameter("commitHash", datasetVersionId);
+        CommitEntity commitEntity = getCommitQuery.uniqueResult();
+        if (commitEntity == null || commitEntity.getParent_commits().isEmpty()) {
+          continue;
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java:434]:
```suggestion
          LOGGER.warn("skipping deleting commit corresponding to dataset version {}", datasetVersionId);
          continue;
```

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java [hunk 3]
```diff
@@ -382,18 +390,122 @@ public CommitEntity getCommitEntity(
...
+        }
+
+        if (commitEntity.getRepository() != null && commitEntity.getRepository().size() > 1) {
+          throw new ModelDBException(
+              "DatasetVersion '"
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java:439]: assert on 425 is point less because of the checks in the if above

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java [hunk 4]
```diff
@@ -382,18 +390,122 @@ public CommitEntity getCommitEntity(
...
-      boolean isDatasetVersion)
+      List<String> datasetVersionIds,
+      RepositoryDAO repositoryDAO)
       throws ModelDBException {
     try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
+      RepositoryEntity repositoryEntity = null;
+      if (repositoryIdentification != null) {
+        repositoryEntity =
+            repositoryDAO.getRepositoryById(session, repositoryIdentification, true, false);
+      }
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java:420]: move this in the else on line 431

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java [hunk 5]
```diff
@@ -382,18 +390,122 @@ public CommitEntity getCommitEntity(
...
+                "DatasetVersion '"
+                    + commitEntity.getCommit_hash()
+                    + "' associated with multiple datasets",
+                Code.INTERNAL);
+          }
+        }
+
+        if (repositoryEntity == null) {
+          repositoryEntity =
+              repositoryDAO.getRepositoryById(session, repositoryIdentification, true, false);
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java:462]: we are callling this twice , we should optimize this

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java [hunk 6]
```diff
@@ -382,18 +390,122 @@ public CommitEntity getCommitEntity(
...
+        }
+
+        Query query = session.createQuery(RepositoryDAORdbImpl.CHECK_BRANCH_IN_REPOSITORY_HQL);
+        query.setParameter("repositoryId", repositoryEntity.getId());
+        query.setParameter("branch", ModelDBConstants.MASTER_BRANCH);
+        BranchEntity branchEntity = (BranchEntity) query.uniqueResult();
+
+        session.beginTransaction();
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java]: we can begin transaction lower

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java [hunk 7]
```diff
@@ -382,18 +390,122 @@ public CommitEntity getCommitEntity(
...
+              SetBranchRequest.newBuilder()
+                  .setRepositoryId(repositoryIdentification)
+                  .setBranch(ModelDBConstants.MASTER_BRANCH)
+                  .setCommitSha(parentDatasetVersion.getCommit_hash())
+                  .build(),
+              false);
+        }
+
+        if (!commitEntity.getChild_commits().isEmpty()) {
+          List<CommitEntity> childCommits = new ArrayList<>(commitEntity.getChild_commits());
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java]:
```suggestion
          CommitEntity childCommits = new ArrayList<>(commitEntity.getChild_commits()).get(0);
```

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java [hunk 8]
```diff
@@ -382,18 +390,122 @@ public CommitEntity getCommitEntity(
...
+          String compositeId =
+              VersioningUtils.getVersioningCompositeId(
+                  repositoryEntity.getId(),
+                  commitEntity.getCommit_hash(),
+                  Collections.singletonList(ModelDBConstants.DEFAULT_VERSIONING_BLOB_LOCATION));
+          DeleteEntitiesCron.deleteLabels(
+              session, compositeId, IDTypeEnum.IDType.VERSIONING_REPO_COMMIT_BLOB);
+          DeleteEntitiesCron.deleteAttribute(session, compositeId);
+          DeleteEntitiesCron.deleteTagEntities(
+              session, repositoryEntity.getId(), commitEntity.getCommit_hash());
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java]: move these to utils , calling them from DEleteEntitiesCron is bad.

Is there a way a dataset repo can have tags? if not then reduce a call

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java [hunk 9]
```diff
@@ -382,18 +390,122 @@ public CommitEntity getCommitEntity(
...
+          session.delete(commitEntity);
+        } else {
+          commitEntity.getRepository().remove(repositoryEntity);
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java]: when will this happen for dataset? throw an exception?

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java [hunk 10]
```diff
@@ -382,18 +390,122 @@ public CommitEntity getCommitEntity(
   }
 
   @Override
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java:409]:
```suggestion
  /**
   * Deleting dataversiosn stored as commits
   * 1. get Repo
   * 2. Iterate through each dataset version
   * 2.1 get Commit
   * 2.2 commit since representing datasetversion verfiy commit belongs to a single repo and match with repo id from 1. 
   * 2.3 get parent commit , since it is dataset assume just single commit
   * 2.4 if commit to be deleted is pointed to by a branch , move branch to parent.
   * 2.5 if commit has children move them to parent
   * 2.5 delete label , tag , attributes for the commit
   */
  @Override
```

**Review by ravishetye — CHANGES_REQUESTED** [review]

File: backend/src/main/java/ai/verta/modeldb/cron_jobs/DeleteEntitiesCron.java (hunk 1 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/cron_jobs/DeleteEntitiesCron.java]: Removed from here and created new PR: https://github.com/VertaAI/modeldb/pull/920

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java (hunk 2 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java:434]: Added

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java (hunk 4 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java:420]: If we moved it to line 431 then it will call in for loop so this is the right place to call it.

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java (hunk 5 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java:462]: No, we are not calling it twice. if `repositoryIdentification` is null this will call otherwise line +401 will call.

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java (hunk 6 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java]: Move it lower before the update operation

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java (hunk 7 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java]: updated as per the suggestion

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java (hunk 8 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java]: `deleteTagEntities` not used in DatasetVersion so removed it from here.

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java (hunk 9 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java]: We are throwing exception from above code `if (commitEntity.getRepository() != null && commitEntity.getRepository().size() > 1) {` so this else never call. i have removed it.

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java (hunk 10 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java:409]: Added

**Review by ravishetye — APPROVED** [review]
